lib/engine/game/g_steam_over_holland/step/buy_train.rb

```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+require_relative '../../../step/buy_train'
+
+module Engine
+  module Game
+    module GSteamOverHolland
+      module Step
+        class BuyTrain < Engine::Step::BuyTrain
+          def actions(entity)
+            return [] if entity != current_entity
+
+            return ['buy_train'] if can_afford_needed_train?(entity)
+            return ['sell_shares'] if entity == current_entity&.owner && can_ebuy_sell_shares?(current_entity)
+            return %w[sell_shares buy_train pass] if president_may_contribute?(entity)
+            return %w[buy_train pass] if can_buy_train?(entity)
+
+            []
+          end
+
+          def pass_description
+            if can_close_corp?(current_entity)
+              "Close #{current_entity.name}"
+            else
+              @acted ? 'Done (Trains)' : 'Skip (Trains)'
+            end
+          end
+
+          def process_pass(action)
+            entity = action.entity
+            return super unless can_close_corp?(entity)
+
+            @game.close_corporation(entity)
+          end
+
+          def must_sell_shares?(corporation)
+            return false if corporation.cash >= @game.depot.min_depot_price || !corporation.trains.empty?
+
+            can_issue?(corporation)
+          end
+
+          def can_issue?(corporation)
+            return false unless corporation.corporation?
+            return false if @round.issued_shares[corporation]
+
+            @game.issuable_shares(corporation).any?
+          end
+
+          def should_buy_train?(entity)
+            :close_corp if can_close_corp?(entity)
+          end
+
+          def can_close_corp?(entity)
+            entity.trains.empty? &&
+              !can_issue?(entity) &&
+              entity.cash + entity.owner.cash < @game.depot.min_depot_price &&
+              @game.graph.route_info(entity)&.dig(:route_train_purchase)
+          end
+
+          def can_afford_needed_train?(entity)
+            entity.trains.empty? &&
+              entity.cash + entity.owner.cash >= @game.depot.min_depot_price &&
+              @game.graph.route_info(entity)&.dig(:route_train_purchase)
```

Comment:
This method is called from `actions` which means it will get called many times when a game is being loaded. Because of this you shouldn't do anything that involves checking the game graph, as that will slow down the game load massively, as it will have to be recalculated multiple times.

You've copied this line from `Engine::Step::Train.must_buy_train?` but that method is called from the game UI, not from the `actions` method. You'll need to work out a way of doing something similar here.

Maybe always give the player the option of passing, but then throw an error in `process_pass` if they still don't have a train and the corporation has a route. There should be something in 1830-type games you can copy.

Reply:
very good point. I think I've done it correctly now. Have a look and let me know.